packages/loopring_v3/contracts/amm/libamm/AmmBlockReceiver.sol

```diff
@@ -21,73 +21,88 @@ library AmmBlockReceiver
     using AmmUpdateProcess  for AmmData.State;
     using BlockReader       for ExchangeData.Block;
 
-    function beforeBlockSubmission(
-        AmmData.State      storage S,
-        ExchangeData.Block memory  _block,
-        bytes              memory  poolTxData,
-        uint                       txIdx
-        )
+    function beforeAllBlocks(AmmData.State storage S)
         internal
-        returns (uint)
-    {
-        AmmData.Context memory ctx = _getContext(S, _block, txIdx);
-
-        BlockReader.BlockHeader memory header = _block.readHeader();
-        require(header.exchange == address(ctx.exchange), "INVALID_EXCHANGE");
-
-        S.approveAmmUpdates(ctx, true);
-
-        _processPoolTx(S, ctx, poolTxData);
-
-        S.approveAmmUpdates(ctx, false);
-
-        // Update state
-        S.poolTokenBurnedSupply = ctx.poolTokenBurnedSupply;
-
-        return ctx.txIdx - txIdx;
-    }
-
-    function _getContext(
-        AmmData.State      storage S,
-        ExchangeData.Block memory  _block,
-        uint                       txIdx
-        )
-        private
         view
         returns (AmmData.Context memory)
     {
         uint size = S.tokens.length;
         return AmmData.Context({
-            _block: _block,
-            txIdx: txIdx,
+            txIdx: 0,
             exchange: S.exchange,
             exchangeDomainSeparator: S.exchange.getDomainSeparator(),
             domainSeparator: S.domainSeparator,
             accountID: S.accountID,
             poolTokenID: S.poolTokenID,
-            poolTokenBurnedSupply: S.poolTokenBurnedSupply,
-            size: size,
             tokens: S.tokens,
-            tokenBalancesL2: new uint96[](size)
+            tokenBalancesL2: new uint96[](size),
+            poolTokenBurnedSupply: S.poolTokenBurnedSupply,
+            pendingTxIdx: 0,
+            pendingTxOwners: new address[](size * 3),
+            pendingTxHashes: new bytes32[](size * 3)
```

Comment:
Seems like the code currently depends on these values to remain updated across calls to `onAmmTransaction`, until `afterAllBlocks` is called once/context object, but the context object is passed into an external/public call, so any updates to the struct in `onAmmTransaction` will be lost after the call ends. The updated context object would need to be returned by  `onAmmTransaction` if it also stores these dynamic variables (or could be stored in a different object, not sure about gas costs).

`pendingTxOwners` and `pendingTxHashes` are only big enough for a single call to `onAmmTransaction`. We should calculate those offchain and pass them onchain (would expensive to calculate onchain).